nncf/quantization/algorithms/post_training/algorithm.py

```diff
@@ -188,28 +211,207 @@ def apply(
+        :return: Updated model with a new submodel.
+        """
+        transformation_layout = TransformationLayout()
+        command = self._backend_entity.create_update_subgraph_command(if_node, if_submodel_condition, submodel)
+        transformation_layout.register(command)
+        return model_transformer.transform(transformation_layout)
+
+    def _add_outputs_before_if_node(
+        self, model_transformer: ModelTransformer, model: TModel, if_node: NNCFNode
+    ) -> TModel:
+        """
+        Inserts extra outputs on If node inputs.
+
+        :param model_transformer: ModelTransformer instance.
+        :param model: Model instance.
+        :param if_node: If node.
+        :return: Model with extra outputs before If node.
+        """
+        transformation_layout = TransformationLayout()
+        for command in self._backend_entity.create_output_insertion_commands(model, if_node):
+            transformation_layout.register(command)
+        return model_transformer.transform(transformation_layout)
+
+    def _dfs_quantize_models(
+        self,
+        parent_model: TModel,
+        parent_graph: NNCFGraph,
+        parent_dataset: Dataset,
+        parent_statistic_points: Optional[StatisticPointsContainer],
+        parent_model_cnt: int,
+    ) -> Tuple[TModel, int]:
+        """
+
+        :param parent_model:
+        :param parent_graph:
+        :param parent_dataset:
+        :param parent_statistic_points:
+        :param parent_model_cnt:
+        :return:
+        """
+        if self._has_if_op(parent_graph, self._backend_entity.if_node_metatype):
+            model_transformer = factory.ModelTransformerFactory.create(parent_model)
+
+            global_model_cnt = parent_model_cnt
+            for if_node in parent_graph.get_nodes_by_metatypes([self._backend_entity.if_node_metatype]):
+                for if_submodel_condition in (True, False):
+                    model_cnt = global_model_cnt + 1
+                    child_model = self._extract_if_submodel(model_transformer, if_node, if_submodel_condition)
+                    child_model_input_names = self._backend_entity.get_if_subgraph_input_names(
+                        parent_model, if_node, if_submodel_condition
+                    )
+                    if_cond_input_name = self._backend_entity.get_if_cond_input_name(parent_model, if_node)
+                    parent_model_with_additional_outputs = self._add_outputs_before_if_node(
+                        model_transformer, parent_model, if_node
+                    )
+
+                    child_dataset = self._make_dataset_for_child_model(
+                        factory.EngineFactory.create(parent_model_with_additional_outputs),
+                        parent_dataset,
+                        if_cond_input_name,
+                        child_model_input_names,
+                        if_submodel_condition,
+                        model_cnt,
+                    )
```

Comment:
Can't we collect the outputs once for all subgraphs? It looks like we can.

Reply:
We can - yes. But should we store all datatitems for all subgraphs at one time?

Reply:
I don't see any reason to increase pipeline time using the loop with the model execution at each step instead of collecting statistics for all models at one time.
Does it really explode memory consumption? Are there any numbers of it based on experiments with a few `if` sub-graphs?